### PR TITLE
MAIN-7204 migrate to typescript, remove gulp

### PR DESCRIPTION
**Pull request by joebowbeerxealth**
Changes:
* Remove extraneous `partner-api` folder
* Update to babel7, with typescript support
* Convert *.js to *.ts using `@khanacademy/flow-to-ts`
* Remove gulp, use babel directly
* Add `@typescript-eslint` (*)
* Add `type-check` and `lint` scripts
* Fix lint errors: `npm run lint -- --fix`

(*) Note: Typescript roadmap is giving preferential treatment to eslint now.

### CODE REVIEW

File: src/auth.ts [hunk 1]
```diff
@@ -19,47 +18,42 @@ export type RequestParam = {
...
-      'Date': dateStr,
-      'Authorization': auth
+      Accept: accept,
+      Date: dateStr,
+      Authorization: auth
     }
   }
 
   if (body) {
-    options = Object.assign(options, { body })
```
Comment (liam-murray-xealth) [inline src/auth.ts]: Maybe use `{...options, body}` (object spread) while at it?

File: src/commands/demo/preorderForms.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  uiSchema: {}
+}
+
+function demo(api: PartnerApi) {
+  return api
+    .putPreorderForm(FORM_ID, ADD_FORM_BODY)
+    .then(() => {
+      return api.getPreorderForm(FORM_ID)
+    })
+    .then(form => {
```
Comment (liam-murray-xealth) [inline src/commands/demo/preorderForms.ts]: Maybe use fat arrow by default consistently while at it (as opposed function() in other places in promise chains).

Comment (joebowbeerxealth) [reply src/commands/demo/preorderForms.ts]: Unfortunately, this file was not recorded as a rename. I barely touched it. Agree that we should improve our sample code!

File: src/auth.ts (hunk 1 above)
Comment (joebowbeerxealth) [reply src/auth.ts]: Agree, though outside scope of typescript conversion.

**Comment by joebowbeerxealth** [comment]
I only fixed the lint errors in this pass. Retained the lint warnings for the time being.

**Pull request merged**